test/core/networking/protocol/authentication_test.dart

```diff
@@ -0,0 +1,94 @@
+import 'package:test/test.dart';
+
+import 'package:flyde/core/networking/protocol/authentication.dart';
+import 'package:flyde/core/networking/protocol/protocol.dart';
+
+void main() {
+  group('AuthRequest', () {
+    test('In ProtocolDelegate.elements', () {
+      expect(ProtocolDelegate.elements['AuthRequest'], equals(isA<Deserialize>()));
+    });
+
+    test('Can serialize', () {
+      final AuthRequest request = AuthRequest(username: "testUser", password: "testPassword");
+
+      final String serialized = ProtocolDelegate.serialize(request);
+
+      expect(
+        serialized,
+        equals(
+          '{"type":"AuthRequest","data":{"username":"testUser","password":"testPassword"}}',
+        ),
+      );
+    });
+
+    test('Can deserialize', () {
+      final String request =
+          '{"type":"AuthRequest","data":{"username":"testUser","password":"testPassword"}}';
+
+      final dynamic deserialized = ProtocolDelegate.deSerialize(request);
```

Comment:
You could consider making `deserialize` generic for automatic type cast

Reply:
We dont know beforehand what we get, so no. I don't want to do: `final deserialized = ProtocolDelegate.deserialize<dynamic>(request);`.

Reply:
In your tests you write `deserialized as Anything` very often, because you know the type. 
If you do not specify a type for the generic it will fallback to dynamic automatically.

Test it out. Paste this to any file in VSCode and hover over `x`. You'll see it's `final dynamic x`
```dart
T get<T>() {
  return null as T;
}

void main() {
  final x = get(); // x is of type dynamic
}
```

Reply:
Do we really need these? The only code that really interacts with the `ProtocolDelegate` is the `protocolMiddleware`, which will never assume any type whatsoever by design. We also don't need this for tests, as testing the correct type is the whole point.

Reply:
Well at least there are so far 12 instances of `deserialized as XXX;` in your code. I'd really like if you could discard these calls. 
Imo they look pretty out of place. If you do not go with the recommended way, maybe you could at least consider writing 

```dart
final MockSerializable deserialized = ProtocolDelegate.deserialize(serialized);
```

that wouldn't change your tests at all and it would be more obvious what you are expecting to be returned.

Reply:
I expect to get a `dynamic` back. After testing its type we can then safely consider it to be type XXX which is when we use `as`. Writing `final MockSerializable deserialized = ProtocolDelegate.deserialize(serialized);` would imply, that `deserialize` definitely returns a `MockSerializable`, which it doesn't. The whole point of these tests is to treat the return as dynamic until we have established the concrete type. I don't see why we should abandon that part of testing.

Reply:
But why do you **want** to get a `dynamic` back? You could as good implement the type casting into `deserialized` and even throw a custom error if the types do not match. Why do you want to make type casting a extra step why has to be repeated whenever someone uses `deserialize`? I think that's a design flaw.

Reply:
It's not, we simply don't know what type we will get back (except in test cases) → `dynamic`. I don't want to implement this only for tests to stay as close to the real use case as possible. The data passed to `deserialize` is not in our control, so we can't make any assumptions.